### ISSUE TITLE
[BUG] index template priority issue causing the server to crash on upgrading from 2.7 to 2.9 or 2.10

### ISSUE THREAD

**Issue by Bogendra-Betapudi** [opensearch-project/observability#1771]
**Describe the bug**
When upgrading the Open search server from 2.7 to 2.9 or 2.10, the server crashes in loop and tries to restart and in the logs we notice an error w.r.to the index templates and priority.
Upon further investigation, found this [thread](https://forum.opensearch.org/t/java-lang-illegalargumentexception-index-template-how-critical/15306/15) where one of the other user also ran into the same issue. But in my case, the server doesn't start up as it tries to restart in a loop and crashes eventually.

**To Reproduce**
1. Have a server installed and running with version prior to 2.9 (mine was 2.7)
2. Now, upgrade the server to 2.9 or 2.10 by downloading the corresponding zip/tar from the official website.
3. Restart the server with the newer version artifacts
4. See error in the logs and the server will not be up for servicing and eventually crash.
5. Below is the snippet of the error that is continuously seen in the logs:

[2023-10-09T14:54:01,342][ERROR][o.o.b.Bootstrap          ] [localhost] Exception
java.lang.IllegalArgumentException: index template **[ss4o_metrics_template]** has index patterns [ss4o_metrics-*-*] matching patterns from existing templates **[ss4o_metric_template]** with patterns (ss4o_metric_template => [ss4o_metrics-*-*]) that have the same priority [1], multiple index templates may not match during index creation, please use a different priority

Note that this is happening even with plain vanilla installation. The steps I tried are here: [https://github.com/opensearch-project/observability/issues/1771](url)

**Expected behavior**
The upgrade should be successful and the server should be up and running without any errors/issues.

**Plugins**
Please list all plugins currently enabled.

**Screenshots**

**Host/Environment (please complete the following information):**
 - OS: Windows

**Additional context**
I tried to get past this by trying to update the priority of the existing template but that resulted in creating a new template instead. This is a blocker as once we try to upgrade, it will fail due to this issue and we can't go back to using the older version as well as the index context has been updated with the latest Open search version.

**Comment by Swiddis** [comment]
The best solution I can find without a patch release is to delete both of the ss4o templates before upgrading. This is safe since the templates are otherwise unused, and avoids any `StartupException`s when upgrading. Might need more detailed reproduction steps if the solution doesn't work.

```sh
# As step 4.5 for the above process
curl -XDELETE "https://opensearch-node1:9200/_index_template/ss4o_*_template"
```

I still wonder why it's intermittently starting successfully, though. In [a linked issue](https://github.com/opensearch-project/OpenSearch/issues/8926) I see both people who see the exception without startup being hindered, and others where it's blocking startup entirely. I'm not sure what differentiates them.

I also have to wonder what tests we need to write to avoid future occurrences, I'd have expected something like this to be revealed by testing.

**Comment by Swiddis** [comment]
Will be resolved in 2.12 onwards following #1770, what's left is verifying a workaround for 2.9-2.11.

**Comment by Bogendra-Betapudi** [comment]
@Swiddis  - Thanks for looking into this. For now, will wait for the 2.12 to be available and would try to use the work around if needed before the release.

**Comment by Swiddis** [comment]
Marking as completed since there's a workaround available and the bug is resolved in 2.12. Workaround: delete the conflicting templates, they're unused/deprecated system templates from Integrations. Specifically the singular forms (`ss4o_trace_template` and `ss4o_metric_template`) are more outdated than the plurals, but all of them are safe to delete.

**Comment by NiceRath** [comment]
As I just ran into this issue - just a note for others:

1. If opensearch fails to start at all (and you have a single-node):
  You can temporarily move the 'state' to get it up again: `mv /var/lib/opensearch/nodes/0/_state /var/lib/opensearch/nodes/0/_state_old`
2. After that you should be able to start it again (in this example on a systemd-based system):
  ```
  systemctl start opensearch.service
  systemctl status opensearch.service
  ```
3. Then delete the templates:
  ```
  # check
  curl -X GET "http://127.0.0.1:9201/_cat/templates"
  # delete
  curl -X DELETE "http://127.0.0.1:9200/_index_template/ss4o_metrics_template"
  curl -X DELETE "http://127.0.0.1:9200/_index_template/ss4o_traces_template"
  ```
4. Not sure if required: Stop opensearch and restore the state:
  `mv /var/lib/opensearch/nodes/0/_state_old /var/lib/opensearch/nodes/0/_state`
5. The service should start-up again
  ```
  systemctl start opensearch.service
  systemctl status opensearch.service
  ```
